### ISSUE TITLE
Can MA0048 support excluded_symbol_names with wildcard?

### ISSUE THREAD

**Issue by abatishchev** [meziantou/Meziantou.Analyzer#602]
**Is your feature request related to a problem? Please describe.**

I'd like to permit only certain classes to be collocated within the same file:

```c#
public class FooHandler {}

public class FooCommand {}
```

or

```c#
public class Bar {}

public static class BarExtensions {}
```

**Describe the solution you'd like**

I tried to adjust the rule using editorconfig;

```
dotnet_diagnostic.MA0048.excluded_symbol_names = *Extensions|*Command|*Query
```

But the build still produces a warning;

```
Warning
MA0048
File name must match type name (https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0048.md)
```

**Describe alternatives you've considered**

None, please suggest if there is any.

**Additional context**

```xml
<PackageReference Include="Meziantou.Analyzer" Version="2.0.87">
  <PrivateAssets>all</PrivateAssets>
  <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
</PackageReference>
```

**Comment by meziantou** [comment]
Would it be ok to only validate the first type in the file? (this is a common request)

**Comment by abatishchev** [comment]
I think something like mode: `all types | first type | disabled` would be great and should cover most of user scenarios.

**Comment by meziantou** [comment]
https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0048.md#configuration

**Comment by abatishchev** [comment]
@meziantou please reopen, I don't see the config taking effect:

Having

```
dotnet_diagnostic.MA0048.only_validate_first_type = true
# which one, btw?
MA0048.only_validate_first_type = true
```

Still produces a warning for a code link this:

```c#
// StartProvisionAccountWorkflowHandler.cs

namespace Handlers.Provisioning
{
    public class ProvisionAccountHandler: IRequestHandler<ProvisionAccountCommand>
    {
    }

    public class ProvisionAccountCommand: ICommand
    {
    }
}
```

**Comment by meziantou** [comment]
According to your code, the warning is still expected.

If the file is named `StartProvisionAccountWorkflowHandler.cs`, the first type of the file must be `StartProvisionAccountWorkflowHandler`, not `ProvisionAccountHandler`

**Comment by abatishchev** [comment]
Sorry, a bad copy & paste. The file is of course named ProvisionAccountHandler.cs:

![image](https://github.com/meziantou/Meziantou.Analyzer/assets/351644/79aa2b20-8a71-4af7-a315-bc165ae17694)

**Comment by meziantou** [comment]
I do not reproduce on my machine

````csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Meziantou.Analyzer" Version="2.0.92">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
  </ItemGroup>

</Project>
````

- `.editorconfig`

````
[*.cs]
MA0048.only_validate_first_type = true
````

- `ProvisionAccountHandler.cs`

````c#
namespace ConsoleApp21;

internal class ProvisionAccountHandler
{
}

internal class Dummy
{
}
````

<img width="352" alt="image" src="https://github.com/meziantou/Meziantou.Analyzer/assets/509220/90df6133-e629-4da7-8945-2533ccb96873">

**Comment by abatishchev** [comment]
How to force the analyzer to reanalyze? After making the changes back and forth, I'm unable to reproduce anything now (VS 2022 versions 17.7.5 and 17.8.0 Preview 2 both misbehave the same way)

**Comment by meziantou** [comment]
The easiest solution is to use the command line: `dotnet build`. Starting with .NET 8, I think you need to use new options: `dotnet build --no-incremental --disable-build-servers`

If you use Visual Studio, update the package, close VS and reopen it. Most of the time this force VS to load the latest version of the analyzer. Then, you can use the command in the menu

<img width="383" alt="image" src="https://github.com/meziantou/Meziantou.Analyzer/assets/509220/d5d35290-1ddc-4c09-9a6e-95f2b8e187e7">

If you use Rider, I there should be a similar way to trigger a full analysis.

**Comment by abatishchev** [comment]
thanks, @meziantou! Here's what I got:

```c#
// ProvisionAccountHandler.cs
namespace ConsoleApp1
{
    internal class ProvisionAccountHandler
    {
    }

    internal class Dummy
    {
    }
}
```

and then this command:

> dotnet build --no-incremental --disable-build-servers

produces this output:

```
MSBuild version 17.8.0-preview-23418-03+0125fc9fb for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Program Files\dotnet\sdk\8.0.100-rc.1.23455.8\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInferenc
e.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-polic
y [D:\Projects\Temp\ConsoleApp1\ConsoleApp1.csproj]
[NuGet Manager] [Warning] The plugin credential provider could not acquire credentials. Authentication may require manual action. Consider re-running the command with --interactive for `dotnet`, /p:NuGetInteractive="true" for MSBuild or removing the -NonInteractive switch for `NuGet`
D:\Projects\Temp\ConsoleApp1\ProvisionAccountHandler.cs(7,20): warning MA0048: File name must match type name (https://
github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0048.md) [D:\Projects\Temp\ConsoleApp1\ConsoleApp1.cspro
j]
  ConsoleApp1 -> D:\Projects\Temp\ConsoleApp1\bin\Debug\net8.0\ConsoleApp1.dll

Build succeeded.

D:\Projects\Temp\ConsoleApp1\ProvisionAccountHandler.cs(7,20): warning MA0048: File name must match type name (https://
github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0048.md) [D:\Projects\Temp\ConsoleApp1\ConsoleApp1.cspro
j]
    1 Warning(s)
    0 Error(s)

Time Elapsed 00:00:04.40
```

Environment:

```
.NET SDK:
 Version:   8.0.100-rc.1.23455.8
 Commit:    e14caf947f

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100-rc.1.23455.8\

.NET workloads installed:
There are no installed workloads to display.

Host:
  Version:      8.0.0-rc.1.23419.4
  Architecture: x64
  Commit:       92959931a3
  RID:          win-x64

.NET SDKs installed:
  6.0.414 [C:\Program Files\dotnet\sdk]
  7.0.401 [C:\Program Files\dotnet\sdk]
  8.0.100-rc.1.23455.8 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.22 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-preview.7.23375.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-rc.1.23421.29 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.22 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-rc.1.23419.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.22 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0-rc.1.23420.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found
```

**Comment by abatishchev** [comment]
Ah, noticed that I downgraded back to `2.0.87` during testing. It works fine in `2.0.92`.

**Comment by abatishchev** [comment]
I wonder why this rule is configurable this way:

```
MA0048.only_validate_first_type = true
```

while others with the `dotnet_diagnostic` prefix, e.g.:

```
dotnet_diagnostic.MA0004.severity = none
```

**Comment by meziantou** [comment]
`dotnet_diagnostic.{id}.severity` is handled by Roslyn directly. If I use the same format I may conflict with new Roslyn properties in the future.

**Comment by abatishchev** [comment]
What would it take to implement the Roslyn-native key `dotnet_diagnostic.MA0048.excluded_symbol_names`?

**Comment by meziantou** [comment]
Roslyn only provides a `string Get("file_path", "key")` method to get a value from the `.editorconfig` (or any configuration system). An analyzer can read any key but you need to handle it yourself. I think Roslyn only reads `dotnet_diagnostic.{id}.severity` as it allows to enable or disable analyzers. All other keys are custom implementation by analyzers. 

So, to support `dotnet_diagnostic.MA0048.excluded_symbol_names`, you need to implement it yourself.

**Comment by abatishchev** [comment]
Would it work if before this line:

https://github.com/meziantou/Meziantou.Analyzer/blob/4e703c6f300c47f80c7f7e74fde37759b9c53a81/src/Meziantou.Analyzer/Rules/FileNameMustMatchTypeNameAnalyzer.cs#L55

Something like this is inserted:

```c#
// copied from the below
var filePath = location.SourceTree.FilePath;
var fileName = filePath == null ? null : GetFileName(filePath.AsSpan());

vat excludedSymbolNames  = context.Options.GetConfigurationValue(location.SourceTree, s_rule.Id + ".excluded_symbol_names", defaultValue: string.Empty);
var excludedSymbolNamesSplit = excludedSymbolNames .Split('|');

// implements exact match
if (excludedSymbolNamesSplit.Contains(fileName))
{
    return;
}

// TODO: implement wildcard

// the rest of the code
```

If yes, I'd be willing to contribute to the project

**Comment by meziantou** [comment]
You can submit a PR that updates the code and the documentation.

Just be sure to avoid useless work and allocations when the configuration is not set (which is the default case).

````
vat excludedSymbolNames = context.Options.GetConfigurationValue(location.SourceTree, s_rule.Id + ".excluded_symbol_names", defaultValue: null);
if(!string.IsNullOrEmtpy(excludedSymbolNames))
{
    ...
}
````

Also, you can check how other analyzer read this kind of configuration. If I well remember there are similar cases.

**Comment by abatishchev** [comment]
Submitted a PR draft, will work on it in the next couple of days. Can you please reopen this issue as it was specifically about `excluded_symbol_names`. Completed PR will close it back.
